File: homework_prompts/rb/w2_d4_trie.rb
#  Homework 11 - Trie
#
#  Problem 1: TrieNode class
#
#  Prompt:    Create a TrieNode class
#             The TrieNode class should contain the following properties:
#
#                 value:   {Character} - default None
#                  next:   {HashTable}
#                   end:   {Boolean}
#
#
#               Example:   sample = TrieNode.new('b')
#                          sample.value     # 'b'
#                          sample.end       # False
#
#  Problem 2:  Trie class
#
#  Prompt:     Create a Trie class
#              The Trie class should contain the following properties:
#
#                  root:   {TrieNode}
#
#              The Trie class should also contain the following methods:
#
#                insert:   A method that adds a word.
#
#                          Input:     word {String}
#                          Output:    {None}
#
#                isWord:   A method that checks whether a word is in the trie.
#
#                          Input:     word {String}
#                          Output:    {Boolean}
#
#              isPrefix:   A method that checks whether an input is a prefix of
#                          a word in the string.
#
#                          Input:     prefix {String}
#                          Output:    {Boolean}
#
#            startsWith:   A method that returns all words that starts with an
#                          input prefix.
#
#                          Input:     prefix {String}
#                          Output:    {List of Strings}
#
#          EXTRA CREDIT:
#                remove:   Removes a word from the trie.
#
#                          Input:     word {String}
#                          Output:    {None}
#


class TrieNode

  def initialize(value = nil)
    # YOUR WORK HERE
  end
    # YOUR WORK HERE
end


class Trie

    def initialize
      # YOUR WORK HERE
    end


    def insert(word)
      # YOUR WORK HERE
    end

    def is_word(word)
      # YOUR WORK HERE
    end


    def starts_with(prefix)
      # YOUR WORK HERE
    end

    def remove(word)
      # YOUR WORK HERE
    end

    # YOUR WORK HERE

end


############################################################
###############  DO NOT TOUCH TEST BELOW!!!  ###############
############################################################

# custom expect function to handle tests
# List count : keeps track out how many tests pass and how many total
#   in the form of a two item array i.e., [0, 0]
# String name : describes the test
# Function test : performs a set of operations and returns a boolean
#   indicating if test passed
def expect(count, name, test)
  count[1] = count[1] + 1

  result = 'false'
  errMsg = nil
  begin
    if test()
      result = ' true'
      count[0] = count[0] + 1
    end
  rescue NameError => e
    MyModule::Logger.error("Error loading the deployer #{class_name}. This deployer it's not installed!")
  rescue StandardError => err
    errMsg = err.message
  end

  puts'  ' + (count[1]).to_s + ')   ' + result + ' : ' + name
  if errMsg != nil
    puts'       ' + errMsg + '\n'
  end
end

# code for capturing puts output
require 'stringio'
require 'ostruct'

class Capture

  def self.capture &block

    # redirect output to StringIO objects
    stdout = StringIO.new
    $stdout = stdout

    result = block.call

    # restore normal output
    $stdout = STDOUT

    OpenStruct.new result: result, stdout: stdout.string
  end
end

# code for checking if lists are equal
def arrays_equal(arr1, arr2)
  if arr1.length != arr2.length
    return false
  end
  for i in 0..arr1.length-1
    if arr1[i] != arr2[i]
      return false
    end
  end
  return true
end

# code for checking if array is sorted (linear runtime)
def is_sorted(input)
  if (input.length < 2)
    return true
  end
  for i in 1..input.length - 1
    if (input[i-1] > input[i])
      return false
    end
  end
  return true
end



puts 'TrieNode Class'
test_count = [0, 0]

def test()
  trieNode = TrieNode.new()
  return trieNode.is_a?(Object)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'able to create an instance', test)

def test()
  trieNode = TrieNode.new()
  return trieNode.instance_variable_defined?(:@value)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has value property', test)

def test()
  trieNode = TrieNode.new()
  return trieNode.instance_variable_defined?(:@value) && trieNode.value == nil
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has default value set to nil', test)

def test()
  trieNode = TrieNode.new()
  if trieNode.instance_variable_defined?(:@value)
    trieNode.value = 'a'
  end
  return trieNode.instance_variable_defined?(:@value) && trieNode.value == 'a'
rescue StandardError => err
  puts err.message
end
expect(test_count, 'able to assign a value upon instantiation', test)

def test()
  trieNode = TrieNode.new()
  trieNode.instance_variable_defined?(:@end)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has an end property', test)

def test()
  trieNode = TrieNode.new()
  return trieNode.instance_variable_defined?(:@end) && trieNode.end == false
rescue StandardError => err
  puts err.message
end
expect(test_count, 'able to assign a value upon instantiation', test)

def test()
  trieNode = TrieNode.new()
  if trieNode.instance_variable_defined?(:@end)
    trieNode.end = true
  end
  return trieNode.instance_variable_defined?(:@end) && trieNode.end == true
rescue StandardError => err
  puts err.message
end
expect(test_count, 'able to assign a value upon instantiation', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"


puts 'Trie Class'
test_count = [0, 0]

def test()
  trie = Trie.new()
  return trie.is_a?(Object)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'able to create an instance', test)

def test()
  trie = Trie.new()
  trie.instance_variable_defined?(:@root)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has root property', test)

def test()
  trie = Trie.new()
  return trie.instance_variable_defined?(:@root) && trie.root.is_a?(TrieNode)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'root property is a TrieNode', test)

def test()
  trie = Trie.new()
  return trie.instance_variable_defined?(:@root) && trie.root.value == nil
rescue StandardError => err
  puts err.message
end
expect(test_count, 'root node value is set to nil', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"


puts 'Trie Insert Method'
test_count = [0, 0]

def test()
  trie = Trie.new()
  trie.respond_to?(:insert)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has insert method', test)

def test()
  trie = Trie.new()
  trie.insert('cat')
  if trie.instance_variable_defined?(:@root)
    c = trie.root.next['c']
    a = c.next['a']
    t = a.next['t']
  end
  return trie.instance_variable_defined?(:@root) && c && a && t && t.end
rescue StandardError => err
  puts err.message
end
expect(test_count, 'able to insert a word into empty trie', test)

def test()
  trie = Trie.new()
  trie.insert('cat')
  trie.insert('cats')
  if trie.instance_variable_defined?(:@root)
    c = trie.root.next['c']
    a = c.next['a']
    t = a.next['t']
    s = t.next['s']
  end
  return trie.instance_variable_defined?(:@root) && c && a && t && t.end && s && s.end
rescue StandardError => err
  puts err.message
end
expect(test_count, 'able to insert words that overlap into trie', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"


puts 'Trie IsWord Method'
test_count = [0, 0]

def test()
  trie = Trie.new()
  trie.respond_to?(:is_word)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has is_word method', test)

def test()
  trie = Trie.new()
  return trie.is_word('') == false
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should return false for an empty string as input', test)

def test()
  trie = Trie.new()
  trie.insert('cat')
  return trie.is_word('cat') == true
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should return true for a word that exists', test)

def test()
  trie = Trie.new()
  trie.insert('cat')
  return trie.is_word('ca') == false
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should return false for a word that does not exist', test)

def test()
  trie = Trie.new()
  trie.insert('cat')
  trie.insert('cats')
  return trie.is_word('cat') == true
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should return true for a word that exists within larger word', test)

def test()
  trie = Trie.new()
  trie.insert('cat')
  trie.insert('cats')
  return trie.is_word('cats') == true
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should return true for a word that is a larger word', test)

def test()
  trie = Trie.new()
  trie.insert('cats')
  return trie.is_word('cat') == false
rescue StandardError => err
  puts err.message
end
expect(test_count, 'should return false if a smaller word was not added, but exists in a larger word', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"


puts 'Trie StartsWith Method'
test_count = [0, 0]

def test()
  trie = Trie.new()
  trie.respond_to?(:starts_with)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has starts_with method', test)

def test()
  trie = Trie.new()
  result = trie.starts_with('cat')
  return result != nil && result.length == 0
rescue StandardError => err
  puts err.message
end
expect(test_count, 'returns an empty array if no words start with input', test)

def test()
  trie = Trie.new()
  trie.insert('cat')
  trie.insert('cats')
  trie.insert('catnip')
  trie.insert('car')
  trie.insert('cars')
  result = trie.starts_with('car')
  return result != nil && result.length == 2 && arrays_equal(result, ['car', 'cars'])
rescue StandardError => err
  puts err.message
end
expect(test_count, 'returns correct prefixes including input that is a word', test)

def test()
  trie = Trie.new()
  trie.insert('cat')
  trie.insert('cats')
  trie.insert('catnip')
  trie.insert('car')
  trie.insert('cars')
  result = trie.starts_with('ca')
  return result != nil && result.length == 5 && arrays_equal(result, ['cat', 'cats', 'catnip', 'car', 'cars'])
rescue StandardError => err
  puts err.message
end
expect(test_count, 'returns the correct prefixes', test)

def test()
  trie = Trie.new()
  trie.insert('cat')
  trie.insert('cats')
  trie.insert('catnip')
  trie.insert('foo')
  trie.insert('hello')
  trie.insert('hell')
  trie.insert('he')
  result = trie.starts_with('')
  return result != nil && result.length == 7 && arrays_equal(result, ['cat', 'cats', 'catnip', 'foo', 'he', 'hell', 'hello'])
rescue StandardError => err
  puts err.message
end
expect(test_count, 'returns all words if input is empty string', test)

puts 'PASSED: ' + (test_count[0]).to_s + ' / ' + (test_count[1]).to_s + "\n\n"


puts 'Trie Remove Method'
test_count = [0, 0]

def test()
  trie = Trie.new()
  trie.respond_to?(:remove)
rescue StandardError => err
  puts err.message
end
expect(test_count, 'has remove method', test)

def test()
  trie = Trie.new()
  trie.insert('cat')
  trie.remove('cat')
  return trie.is_word('cat') == false && trie.root.next['c'] == nil
rescue StandardError => err
  puts err.message
end
expect(test_count, 'removes a word that exists', test)

def test()
  trie = Trie.new()
  trie.insert('cat')
  trie.remove('c')
  return trie.is_word('cat') == true
rescue StandardError => err
  puts err.message
end
expect(test_count, 'does not remove a word that does not exist', test)

def test()
  trie = Trie.new()
  trie.insert('hello')
  trie.insert('hell')
  trie.insert('he')
  trie.remove('hell')
  return trie.is_word('he') == true && trie.is_word('hello') == true && trie.is_word('hell') == false
rescue StandardError => err
  puts err.message
end
expect(test_count, 'does not remove letters that belong to a longer word', test)

def test()
  trie = Trie.new()
  trie.insert('cat')
  trie.insert('cats')
  trie.insert('catnip')
  trie.remove('catnip')
  return trie.is_word('cat') == true && trie.is_word('cats') == true && trie.is_word('catnip') == false && trie.root.next['c'].next['a'].next['t'].next['n'] == nil
rescue StandardError => err
  puts err.message
end
expect(test_count, 'removes letters from longer word and keeps shorter letters', test)

File: homework_prompts/swift/w2_d4_trie.swift
/**
 *  Homework 11 - Trie
 *
 *  Problem 1: TrieNode class
 *
 *  Prompt:    Create a TrieNode class
 *             The TrieNode class should contain the following properties:
 *
 *                 value:   {Character} - default null
 *                  next:   {HashTable}
 *                   end:   {Boolean}
 *
 *
 *               Example:   let sample = new TrieNode("b")
 *                          sample.value     // "b"
 *                          sample.end       // false
 *
 *  Problem 2:  Trie class
 *
 *  Prompt:     Create a Trie class
 *              The Trie class should contain the following properties:
 *
 *                  root:   {TrieNode}
 *
 *              The Trie class should also contain the following methods:
 *
 *                insert:   A method that adds a word.
 *
 *                          Input:     word {String}
 *                          Output:    {undefined}
 *
 *                isWord:   A method that checks whether a word is in the trie.
 *
 *                          Input:     word {String}
 *                          Output:    {Boolean}
 *
 *              isPrefix:   A method that checks whether an input is a prefix of
 *                          a word in the string.
 *
 *                          Input:     word {String}
 *                          Output:    {Boolean}
 *
 *            startsWith:   A method that returns all words that starts with an
 *                          input word.
 *
 *                          Input:     word {String}
 *                          Output:    {Array of Strings}
 *
 *          EXTRA CREDIT:
 *                remove:   Removes a word from the trie.
 *
 *                          Input:     word {String}
 *                          Output:    {undefined}
 */


 class TrieNode {
   var value : Character
   var next : [Character:TrieNode]
   var end: Bool

   init(_ value: Character) {
     self.value = value
     self.next = [:]
     self.end = false
   }
 }

 class Trie {
   var root : TrieNode
   init() {
     self.root = TrieNode(" ")
   }

   func insert(_ word: String) {
     // YOUR WORK HERE
   }

   func isWord(_ word: String) -> Bool {
     // YOUR WORK HERE
     return false
   }

   func isPrefix(_ word: String) -> Bool {
     // YOUR WORK HERE
     return false
   }

   func startsWith(_ word: String) -> [String] {
     // YOUR WORK HERE
     return []
   }

   func remove(_ word: String) {
     // YOUR WORK HERE
   }


 }

 ////////////////////////////////////////////////////////////
 ///////////////  DO NOT TOUCH TEST BELOW!!!  ///////////////
 ////////////////////////////////////////////////////////////




 func assert(_ count: inout [Int], _ name: String, _ test: () -> Bool) {
   if count.count != 2 {
     count = [0, 0]
   } else {
     count[1] = count[1] + 1
   }

   var pass: String = "false"

   if test() {
     pass = " true"
     count[0] = count[0] + 1
   }
   print(count[1], ")   ", pass, ":", name)
 }


 print("TrieNode Class")
 var testCount = [0, 0]


 assert(&testCount, "has value property that is a Character",  {
   let trieNode = TrieNode(" ")
   return trieNode.value == " "
 })

 assert(&testCount, "has end property, defaults to false",  {
   let trieNode = TrieNode(" ")
   return trieNode.end == false
 })

 assert(&testCount, "has next property that is a HashTable",  {
   let trieNode = TrieNode(" ")
   return trieNode.next.count == 0
 })

 assert(&testCount, "able to assign value upon instantiation",  {
   let trieNode = TrieNode(" ")
   trieNode.value = "a"
   return trieNode.value == "a"
 })

 assert(&testCount, "able to assign end upon instantiation",  {
   let trieNode = TrieNode(" ")
   trieNode.end = true
   return trieNode.end == true
 })

 print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")


 print("Trie Class")
 testCount = [0, 0]


 assert(&testCount, "has root property set to empty TrieNode",  {
   let trie = Trie()
   return trie.root.value == " "
 })

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")


print("Trie Insert Method")
testCount = [0, 0]


assert(&testCount, "able to insert a word into empty Trie",  {
  let trie = Trie()
  trie.insert("cat")
  return trie.root.next["c"] != nil && trie.root.next["c"]!.next["a"] != nil && trie.root.next["c"]!.next["a"]!.next["t"] != nil &&
  trie.root.next["c"]!.next["a"]!.next["t"]!.end == true
})

assert(&testCount, "able to insert words that don't overlap",  {
  let trie = Trie()
  trie.insert("cat")
  trie.insert("dog")

  return trie.root.next["c"] != nil && trie.root.next["c"]!.next["a"] != nil && trie.root.next["c"]!.next["a"]!.next["t"] != nil &&
  trie.root.next["c"]!.next["a"]!.next["t"]!.end == true && trie.root.next["d"] != nil && trie.root.next["d"]!.next["o"] != nil && trie.root.next["d"]!.next["o"]!.next["g"] != nil &&
  trie.root.next["d"]!.next["o"]!.next["g"]!.end == true
})

assert(&testCount, "able to insert words that overlap",  {
  let trie = Trie()
  trie.insert("cat")
  trie.insert("cats")
  return trie.root.next["c"] != nil && trie.root.next["c"]!.next["a"] != nil && trie.root.next["c"]!.next["a"]!.next["t"] != nil &&
  trie.root.next["c"]!.next["a"]!.next["t"]!.end == true && trie.root.next["c"]!.next["a"]!.next["t"]!.next["s"] != nil
  && trie.root.next["c"]!.next["a"]!.next["t"]!.next["s"]!.end == true
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")


print("Trie IsWord Method")
testCount = [0, 0]

assert(&testCount, "should return false for an empty string as input",  {
  let trie = Trie()
  return trie.isWord("") == false
})

assert(&testCount, "should return true for a word that exists",  {
  let trie = Trie()
  trie.insert("cat")
  return trie.isWord("cat") == true
})

assert(&testCount, "should return false for a word that exists",  {
  let trie = Trie()
  trie.insert("cat")
  return trie.isWord("dog") == false
})

assert(&testCount, "should return false for a prefix that exists in a larger word",  {
  let trie = Trie()
  trie.insert("cat")
  return trie.isWord("ca") == false
})

assert(&testCount, "should return true for a word that exists in a larger word",  {
  let trie = Trie()
  trie.insert("cat")
  trie.insert("cats")
  return trie.isWord("cat") == true
})

assert(&testCount, "should return true for a larger word that contains smaller words",  {
  let trie = Trie()
  trie.insert("cat")
  trie.insert("cats")
  return trie.isWord("cats") == true
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")


print("Trie StartsWith Method")
testCount = [0, 0]

assert(&testCount, "returns an empty array if no words start with input",  {
  let trie = Trie()
  return trie.startsWith("") == []
})

assert(&testCount, "returns only words that begin with input as prefix",  {
  let trie = Trie()
  trie.insert("cat")
  trie.insert("car")
  trie.insert("can")
  trie.insert("cut")
  trie.insert("dog")
  trie.insert("deer")
  trie.insert("bird")
  let result : [String] = trie.startsWith("c")
  return result.count == 4 && result.contains("cat") && result.contains("car") && result.contains("can") && result.contains("cut")
})

assert(&testCount, "returns result with input, if input is a word",  {
  let trie = Trie()
  trie.insert("cat")
  trie.insert("cats")
  trie.insert("car")
  trie.insert("can")
  trie.insert("cut")
  trie.insert("dog")
  trie.insert("deer")
  trie.insert("bird")
  let result : [String] = trie.startsWith("cat")
  return result.count == 2 && result.contains("cat") && result.contains("cats")
})

assert(&testCount, "returns all words if input is empty string",  {
  let trie = Trie()
  trie.insert("cat")
  trie.insert("car")
  trie.insert("can")
  trie.insert("cut")
  trie.insert("dog")
  trie.insert("deer")
  trie.insert("bird")
  return trie.startsWith("").count == 7
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")


print("Trie Remove Method")
testCount = [0, 0]

assert(&testCount, "removes word that exists",  {
  let trie = Trie()
  trie.insert("cat")
  trie.remove("cat")
  return trie.isWord("cat") == false
})

assert(&testCount, "does not remove word that does not exist",  {
  let trie = Trie()
  trie.insert("cat")
  trie.remove("car")
  return trie.isWord("cat") == true
})

assert(&testCount, "does not remove letters that belong to a longer word",  {
  let trie = Trie()
  trie.insert("cat")
  trie.insert("cats")
  trie.remove("cat")
  return trie.isWord("cats") == true && trie.isWord("cat") == false
})

assert(&testCount, "removes letters from longer word and keeps letters from shorter word",  {
  let trie = Trie()
  trie.insert("cat")
  trie.insert("cats")
  trie.remove("cats")
  return trie.isWord("cat") == true && trie.isWord("cats") == false
})

print("PASSED: ", testCount[0], " / ", testCount[1], "\n\n")
